### PR TITLE
Update to Node 20

### PR DESCRIPTION
**Pull request by jtpio**
## References

Follow-up to #15969.

## Code changes

- Update `@types/node` to version 20
- Update `nodejs` in the `engines` section of `dev_mode/package.json`

## User-facing changes

None

## Backwards-incompatible changes

None

### CODE REVIEW

**Comment by jtpio** [comment]
Pending a release of `jupyterlab/maintainer-tools/.github/actions/base-setup@v1` which includes https://github.com/jupyterlab/maintainer-tools/pull/222.

**Comment by jtpio** [comment]
Failures seem to be coming from Yarn:

```
node_modules/@yarnpkg/fslib/lib/ZipFS.d.ts:206:5 - error TS2416: Property 'readdirPromise' in type 'ZipFS' is not assignable to the same property in base type 'BasePortableFakeFS'.
  Type '{ (p: import("/workspace/jupyterlab/node_modules/@yarnpkg/fslib/lib/path").PortablePath): Promise<import("/workspace/jupyterlab/node_modules/@yarnpkg/fslib/lib/path").Filename[]>; (p: import("/workspace/jupyterlab/node_modules/@yarnpkg/fslib/lib/path").PortablePath, opts: { ...; }): Promise<...>; (p: import("/worksp...' is not assignable to type '{ (p: import("/workspace/jupyterlab/node_modules/@yarnpkg/fslib/lib/path").PortablePath): Promise<import("/workspace/jupyterlab/node_modules/@yarnpkg/fslib/lib/path").Filename[]>; (p: import("/workspace/jupyterlab/node_modules/@yarnpkg/fslib/lib/path").PortablePath, opts: { ...; }): Promise<...>; (p: import("/worksp...'. Two different types with this name exist, but they are unrelated.
    Type 'Promise<Filename[]>' is not assignable to type 'Promise<Dirent[]>'.
      Type 'Filename[]' is not assignable to type 'Dirent[]'.
        Type 'Filename' is not assignable to type 'Dirent'.
          Type 'String & { __pathType: PathType.File; }' is missing the following properties from type 'Dirent': isFile, isDirectory, isBlockDevice, isCharacterDevice, and 5 more.

206     readdirPromise(p: PortablePath): Promise<Array<Filename>>;
        ~~~~~~~~~~~~~~
```

**Comment by jtpio** [comment]
This UI test seems to be failing consistently: 

image

```
Error: locator.waitFor: Error: strict mode violation: locator('.jp-RunningSessions-item.jp-mod-kernel').locator('text="Python 3 (ipykernel)"') resolved to 2 elements:
    1) <span class="jp-RunningSessions-itemLabel" title="Pyth…>Python 3 (ipykernel)</span> aka getByTitle('Python 3 (ipykernel): b07bb759-c807-4bab-8433-fecd0cb1105b')
    2) <span class="jp-RunningSessions-itemLabel" title="Pyth…>Python 3 (ipykernel)</span> aka getByTitle('Python 3 (ipykernel): 50dbe483-bdf8-4c34-ba23-0ca7ab7d0679\n\nnotebook\nPath: notebooks/Data.ipynb')

=========================== logs ===========================
waiting for locator('.jp-RunningSessions-item.jp-mod-kernel').locator('text="Python 3 (ipykernel)"') to be visible
============================================================

  36 |         '.jp-RunningSessions-item.jp-mod-kernel >> text="Python 3 (ipykernel)"'
  37 |       )
> 38 |       .waitFor();
     |        ^
  39 |     expect(
  40 |       await page.screenshot({ clip: { y: 27, x: 0, width: 283, height: 400 } })
  41 |     ).toMatchSnapshot('interface_tabs.png');

    at /home/runner/work/jupyterlab/jupyterlab/core/galata/test/documentation/overview.test.ts:38:8
```

Although not sure it's related to the PR.

**Comment by jtpio** [comment]
> Although not sure it's related to the PR.

There appears to be the same failure on another recent PR: https://github.com/jupyterlab/jupyterlab/pull/16032 -> https://github.com/jupyterlab/jupyterlab/actions/runs/8362598554/job/22893562907?pr=16032

So marking this one as ready.

File: buildutils/package.json [hunk 1]
```diff
@@ -41,8 +41,8 @@
     "watch": "tsc -w --listEmittedFiles"
   },
   "dependencies": {
-    "@yarnpkg/core": "^3.0.0",
-    "@yarnpkg/parsers": "^2.0.0",
+    "@yarnpkg/core": "^4.0.3",
```
[the remark this answers is not in the document]
Comment (krassowski) [reply buildutils/package.json:44]: So these two are only used in `update-staging-lock.ts`. It uses `structUtils.parseDescriptor` and `structUtils.parseRange` which did not have any substantial changes recently. `parseSyml` had not changes and `stringifySyml` had minor changes (https://github.com/yarnpkg/berry/pull/5548/files).

**Review by krassowski — APPROVED** [review]
Thank you!

File: buildutils/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtpio) [reply buildutils/package.json:44]: For reference bumping the packages fixes the TS issue reported here: https://github.com/yarnpkg/berry/issues/5728

And it looks like it was not backported to versions we were using, but thankfully updating to the latest major versions does not require us to update the code.

**Pull request merged**